File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode helper(int[] preorder, int pstart, int pend, int[] inorder, int istart, int iend)
    {
        if (pstart == pend) return null;

        TreeNode root = new TreeNode(preorder[pstart]);
        int idx = istart;
        int leftsz = 0;
        while(inorder[idx] != root.val)
        {
            idx++;
            leftsz++;
        }
        root.left = helper(preorder, pstart + 1, pstart + 1 + leftsz, inorder, istart, idx);
        root.right = helper(preorder, pstart + 1 + leftsz, pend, inorder, idx + 1, iend);
        return root;
    }

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return helper(preorder, 0, preorder.length, inorder, 0, inorder.length);
    }
}

File: Data Structures & Algorithms/house-robber-iii/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        cache = {}
        def traverse(node):
            if node in cache:
                return cache[node]
            if node == None:
                return 0
        
            cache[node] = node.val
            if node.left:
                cache[node] += (traverse(node.left.right) + traverse(node.left.left))
            if node.right:
                cache[node] += (traverse(node.right.right) + traverse(node.right.left))
            cache[node] = max(cache[node], traverse(node.left) + traverse(node.right))
            return cache[node]
        return traverse(root)